### DISCUSSION TITLE
Visibility modifier for [ObservableProperty] Source Generator

### DISCUSSION THREAD

**Discussion by ewerspej** [CommunityToolkit/dotnet#666] [Ideas]
I'm currently converting some ViewModels and their properties of a Xamarin.Forms app to use the Source Generators of the MVVM Toolkit.

I've noticed that there is currently no way to modify the visibility of the property setter of a generated property. There are sometimes situations where you would mark a setter as `internal`,  `private` or `protected` (or even `internal protected`), so that a property can only be set by certain actors, while the getter should always be `public`, so that it can be used in binding expressions.

For example, I would like to be able to use the `[ObservableProperty]` Source Generator to be able to generate the equivalent of the following:

```c#
private string identifier;
public string Identifier
{
    get => identifier;
    private set
    {
        if(identifier.Equals(value)) return;
        identifier = value;
        OnPropertyChanged();
    }
}
```

The `Identifier` property could then still be used in a binding expression:

```xml
<Label Text="{Binding Identifier}" />
```

The value can only be set from inside the class, e.g. in a method or command that calls some API:

```c#
[RelayCommand]
private async Task UpdateAsync()
{
    Identifier = await RestClient.SomeUpdateCallAsync();
}
```

This is useful, for example, if the value of the `Identifier` comes from an API call and shouldn't be publicly settable, or when it should not participate in Two-Way bindings. Therefore there should be a way to modify the visibility.

My proposal is to add an optional visibility modifier to the `[ObservableProperty]` Source Generator called `SetterVisibility` which takes a single argument, e.g. an enum value or some other limited form of identifier:

```c#
//generates property MyBoolA with private setter
[ObservableProperty(SetterVisibility = Visibility.Private)]
private bool myBoolA;

//generates property MyBoolB with internal setter
[ObservableProperty(SetterVisibility = Visibility.Internal)]
private bool myBoolB;

//generates property MyBoolC with protected setter
[ObservableProperty(SetterVisibility = Visibility.Protected)]
private bool myBoolC;
```

The resulting properties should then be equivalent to the following:

```c#
private bool myBoolA;
public bool MyBoolA
{
    get => myBoolA;
    private set
    {
        if(myBoolA.Equals(value)) return;
        myBoolA = value;
        OnPropertyChanged();
    }
}

private bool myBoolB;
public bool MyBoolB
{
    get => myBoolB;
    internal set
    {
        if(myBoolB.Equals(value)) return;
        myBoolB = value;
        OnPropertyChanged();
    }
}

private bool myBoolC;
public bool MyBoolC
{
    get => myBoolC;
    protected set
    {
        if(myBoolC.Equals(value)) return;
        myBoolC = value;
        OnPropertyChanged();
    }
}
```

The default must, of course, be `Visibility.Public` or `null` in order to not break existing behavior.

This would be highly useful and would allow the MVVM Source Generators to cover more scenarios that already were possible before.

**Comment by konstitech** [comment]
The feature would be very useful.  I miss it

**Comment by Sergio0694** [comment]
This is included as part of #555. Closing as duplicate of #291.
